clarity-ext-scripts/clarity_ext_scripts/covid/rtpcr_analysis_service.py

```diff
@@ -26,82 +33,121 @@ class NegativeControlWasPositive(AnalysisServiceException):
-# Positive if CT(FAM) = <=38
-# Failed  if CT(FAM)= >38
-# Negative if CT(FAM) = 0 and CT(HEX) = <=32
-# Retest if CT(FAM) = 0 and CT(HEX) = >32
+    # These are notes on what Maike said about the analysis
+    # Understanding of the analysis criterias
+    # FAM = COVID, HEX = Human internal control
+    # Positive if CT(FAM) = <=38
+    # Failed  if CT(FAM)= >38
+    # Negative if CT(FAM) = 0 and CT(HEX) = <=32
+    # Failed if CT(FAM) = 0 and CT(HEX) = >32
 
-# TODO Should we issue a re-test status here
-# TODO Later we might have to account for giving two trails over 38 as positive.
+    # TODO Later we might have to account for giving two trails over 38 as positive.
 
     def _analyze_sample(self, sample):
-        fam_ct = sample["FAM-CT"]
-        hex_ct = sample["HEX-CT"]
-        if fam_ct > self.FAM_THRESHOLD:
-            return COVID_RESPONSE_FAILED
-        elif fam_ct == 0 and hex_ct <= self.HEX_THRESHOLD:
+        covid_ct = sample["{}-CT".format(self._covid_reporter_key)]
+        internal_control_ct = sample["{}-CT".format(
+            self._internal_control_reporter)]
+        if covid_ct > self.COVID_CONTROL_THRESHOLD:
+            return FAILED_BY_TO_HIGH_COVID_VALUE
+        elif covid_ct == 0 and internal_control_ct <= self.INTERNAL_CONTROL_THRESHOLD:
             return COVID_RESPONSE_NEGATIVE
-        elif fam_ct == 0 and hex_ct > self.HEX_THRESHOLD:
-            # TODO Should retest and failed be the same thing or not.
-            return COVID_RESPONSE_FAILED
-        elif fam_ct <= self.FAM_THRESHOLD:
+        elif covid_ct == 0 and internal_control_ct > self.INTERNAL_CONTROL_THRESHOLD:
+            return FAILED_BY_INTERNAL_CONTROL
+        elif covid_ct <= self.COVID_CONTROL_THRESHOLD:
             return COVID_RESPONSE_POSITIVE
         else:
-            # TODO Don't know if it is actually a good idea to raise here or not...
-            raise NotImplementedError(
-                "Got CT-value for FAM: {} and CT-value for HEX: {}.".format(fam_ct, hex_ct))
+            raise AssertionError(
+                "Got CT-value for {}: {} and CT-value for {}: {}.".format(self._covid_reporter_key,
+                                                                          covid_ct,
+                                                                          self._internal_control_reporter_key,
+                                                                          internal_control_ct))
 
     def _analyze_controls(self, positive_controls, negative_controls):
         errors = []
-        for pos_control in positive_controls:
-            if self._analyze_sample(pos_control) == COVID_RESPONSE_NEGATIVE:
-                errors += PositiveControlWasNegative(
-                    "Positive control sample: {} was negative for covid-19".format(pos_control["id"]))
+        control_results = []
 
+        for pos_control in positive_controls:
+            res = self._analyze_sample(pos_control)
```

Comment:
I'm torn here between if we should care about DRY or not between these two loops. On the one hand it's perhaps most readable to not keep it DRY, on the other hand, they are very similar and their body could be refactored into a function that takes "pos" or "neg" as a parameter.

Reply:
Yeah, I agree. I was a bit torn about this when I wrote it. But for now I think we can stick with this.